video/out/drm_atomic.c

```diff
@@ -452,3 +453,114 @@ bool drm_mode_destroy_blob(int fd, struct drm_mode *mode)
+    switch(color->gamma) {
+        case MP_CSP_TRC_BT_1886: p->data.hdmi_metadata_type1.eotf = 1; break; // ???
+        case MP_CSP_TRC_PQ: p->data.hdmi_metadata_type1.eotf = 2; break;
+        case MP_CSP_TRC_HLG: p->data.hdmi_metadata_type1.eotf = 3; break;
+        default: p->data.hdmi_metadata_type1.eotf = 0; break;		// default SDR
+     }
+     
+
+    if(p->data.hdmi_metadata_type1.eotf) {
+
+        float min_luminance = 0;	// ??? Is it correct?
+        float max_luminance = color->sig_peak / 100.0; // [0..1]; mpv makes color management, so, all levels are the same?
+        float max_cll = color->sig_peak / 100.0;
+        float max_fall = color->sig_peak / 100.0;
+        
+        // Convert to  "16-bit values in units of 0.00002, where 0x0000 represents zero and 0xC350 represents 1.0000"
+        p->data.hdmi_metadata_type1.display_primaries[0].x = round(colors.red.x * 50000.0);
+        p->data.hdmi_metadata_type1.display_primaries[0].y = round(colors.red.y * 50000.0);
+        p->data.hdmi_metadata_type1.display_primaries[1].x = round(colors.green.x * 50000.0);
+        p->data.hdmi_metadata_type1.display_primaries[1].y = round(colors.green.y * 50000.0);
+        p->data.hdmi_metadata_type1.display_primaries[2].x = round(colors.blue.x * 50000.0);
+        p->data.hdmi_metadata_type1.display_primaries[2].y = round(colors.blue.y * 50000.0);
+        // White point
+        p->data.hdmi_metadata_type1.white_point.x = round(colors.white.x * 50000.0);
+        p->data.hdmi_metadata_type1.white_point.y = round(colors.white.y * 50000.0);
+
+        //... "16-bit value in units of 1 cd/m2, where 0x0001 represents 1 cd/m2 and 0xFFFF represents 65535 cd/m2"
+        if(max_luminance >= 0)
+            p->data.hdmi_metadata_type1.max_display_mastering_luminance = round(max_luminance * 65535);
+        //... "16-bit value in units of 0.0001 cd/m2, where 0x0001 represents 0.0001 cd/m2 and 0xFFFF represents 6.5535 cd/m2"
+        if(min_luminance >= 0)
+            p->data.hdmi_metadata_type1.min_display_mastering_luminance = round(min_luminance * 65535);
+        //... "16-bit value in units of 1 cd/m2, where 0x0001 represents 1 cd/m2 and 0xFFFF represents 65535 cd/m2."
+        if (max_cll >= 0)
+            p->data.hdmi_metadata_type1.max_cll = round(max_cll * 65535);
+        //... "16-bit value in units of 1 cd/m2, where 0x0001 represents 1 cd/m2 and 0xFFFF represents 65535 cd/m2."
+        if (max_fall >= 0)
+            p->data.hdmi_metadata_type1.max_fall = round(max_fall * 65535);
+       
+       drmModeCreatePropertyBlob(ctx->connector->fd, &p->data, sizeof(p->data), &p->blob_id);
+       drmModeAtomicReqPtr request = drmModeAtomicAlloc();
+       if (!request)
+            return;
+       drmModeAtomicAddProperty(request, ctx->connector->id, property_hdr_id, p->blob_id);
+       drmModeAtomicCommit(ctx->connector->fd, request, DRM_MODE_ATOMIC_ALLOW_MODESET, NULL);
```

Comment:
The whole point of atomic commits is to submit multiple changes together, _atomically_. Updating one property at a time is an API mis-use, will schedule a new page-flip, and will make any subsequent atomic commit fail with `EBUSY` until the page-flip is complete.

Adiditonally, the return value of `drmModeAtomicCommit` should really be checked here.

Reply:
One can use the TEST_ONLY flag here and track the return status. On success they can pass the blob to the caller, which will add it to the existing state and issue it's `drmModeAtomicCommit`